{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1><center>Data Acquisition Algorithms</center></h1>\n",
    "\n",
    "This notebook contains algorithms for obtaining, aggregating and cleaning miRNA-Seq data for a specific tumor type from the GDC API. Data acquired here is to be used for differential expression analysis.\n",
    "\n",
    "A lot of the code presented here is similar to DataRetrieval, however it is linked together to run all at once."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Packages\n",
    "\n",
    "This notebook will be using multiple packages to query and interpret data. Outlined below is the reason for each.\n",
    "\n",
    "| Package | Use |\n",
    "| --- | --- |\n",
    "|requests| retrieving information from the url endpoints given on the <br> [GDC API User Guide](https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/) |\n",
    "|pandas (pd) | storage and wrangling of data in pandas.DataFrame objects |\n",
    "| json | converting the request responses into JSON format for easy use |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieve File Information/metadata\n",
    "\n",
    "The first step is to retrieve the information on the files pertaining to our anaysis. This information will include file UUIDs that can be used to query the GDC API for the txt file contents. To retrieve UUIDs, the **files** endpoint will be used and multiple filters will be passed, shortcutting to the **cases** endpoint for more selective filtering. The files are then used to create a dictionary that containing the UUID and sample type.\n",
    "\n",
    "\n",
    "\n",
    "#### The Following Cells:\n",
    "\n",
    "\n",
    "* Use the **files** endpoint\n",
    "\n",
    "* Shortcut to the **cases** endpoint\n",
    "\n",
    "* Utilize multiple filters: \"files.experimental_strategy\" and \"cases.project.project_id\"\n",
    "\n",
    "* Return relevant file info in JSON format\n",
    "\n",
    "* Extract important information and save to dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### file_filt()\n",
    "\n",
    "This algorithm is the first step for data acquisition: extracting specific file information from the GDC with particular filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def file_filt(projects):\n",
    "    \"\"\" Algorithm that uses the GDC API to find files that fit our criteria\n",
    "    \n",
    "    Parameters:\n",
    "        projects: list of comma seterpated project names coded as strings\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    url=\"https://api.gdc.cancer.gov/files\"                              # url found on the GDI website\n",
    "\n",
    "    filt = {                                                            # creates dictionary of filtering parameters\n",
    "        \"op\":\"and\",\n",
    "        \"content\":[\n",
    "            {\n",
    "                \"op\":\"=\",                                               # filters for miRNA expression Quantification..\n",
    "                \"content\":{                                             # .. which increases speed of other filtering\n",
    "                    \"field\":\"files.data_type\",\n",
    "                    \"value\":[\"miRNA Expression Quantification\"]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "                \"op\":\"=\",                                               # filter for miRNA-seq information\n",
    "                \"content\":{\n",
    "                    \"field\":\"files.experimental_strategy\",\n",
    "                    \"value\":[\"miRNA-Seq\"]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "                \"op\":\"in\",                                              # filter by Project title, this takes a comma seperated..\n",
    "                \"content\":{                                             # .. list, allowing for us to input specific projects\n",
    "                    \"field\":\"cases.project.project_id\",\n",
    "                    \"value\":projects\n",
    "                }\n",
    "            }\n",
    "                ]\n",
    "    }\n",
    "\n",
    "    D = {\"filters\":json.dumps(filt),\n",
    "        \"size\":\"2500\",\n",
    "        \"expand\": \"cases.project\",\n",
    "        \"fields\":\"file_id,file_name,cases.submitter_id,cases.samples.sample_type,data_format\"}\n",
    "\n",
    "    r = requests.get(url, params=D) \n",
    "    \n",
    "    cats=r.json()\n",
    "    \n",
    "    return cats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### file_Dict()\n",
    "\n",
    "This algorithm uses the json object returned by **file_filt()** and creates a dictionary of UUIDs and sample types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def file_Dict(cats):\n",
    "    \"\"\" Creats a dictionary from the output of file_filt() containing id:sample_type\n",
    "    \n",
    "    parameters:\n",
    "        cats: json of all information outputted from the file_filt() agorithm\n",
    "       \n",
    "       returns a dictionary\n",
    "    \"\"\"\n",
    "    a={}\n",
    "    for i in range(len(cats[\"data\"][\"hits\"])):                           # iterates over all files and counts different sample..\n",
    "                                                                         # type\n",
    "        a[cats[\"data\"][\"hits\"][i]['file_id']]=cats[\"data\"][\"hits\"][i]['cases'][0]['samples'][0][\"sample_type\"]\n",
    "    \n",
    "    return a"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieve Data from list above\n",
    "\n",
    "The next step is to use the UUIDs from the previous query to retrieve file data.\n",
    "\n",
    "#### The Following Cells:\n",
    "\n",
    "- Use the **data** endpoint\n",
    "\n",
    "- Use a simple HTTP GET request including the file UUID (file_id)\n",
    "\n",
    "\n",
    "__Note:__ According to the GDC API Users Guide, retrieving multiple files requires a POST request. However, due to the nature of our data (relatively small size) I do not think this is necessary. I believe POST requests are required for genomic datasets for easier acquisistion without large compuational power."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### data_retrieval()\n",
    "\n",
    "This algorithm does a GET request to the GDC API using a file UUID acquired earlier. This algorthim will exist nestled in a for-loop that iterates over the dictionary created from **file_Dict()**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def data_retrieval(UUID):\n",
    "    \"\"\" retrieves data through the GDC API from a UUID\n",
    "    \n",
    "    Parameters:\n",
    "        UUID: file_id\n",
    "    \n",
    "    returns content of the return\n",
    "    \"\"\"\n",
    "    \n",
    "    url = 'https://api.gdc.cancer.gov/data/'+ UUID     # Add string of file UUID to url.\n",
    "\n",
    "    D = {}\n",
    "\n",
    "    r = requests.get(url, params=D, headers = {\"Content-Type\": \"application/json\"})\n",
    "    \n",
    "    return r.content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean and Convert the HHTP response content into a DataFrame\n",
    "\n",
    "The contents of the request reponse come in byte format. To use the data retrieved, it must be converted to a string and each row saved as a DataFrame. The *Pandas* package will be used to store and wrangle data.\n",
    "\n",
    "\n",
    "#### The Following Cells:\n",
    "\n",
    "- Use *Pandas* DataFrames\n",
    "\n",
    "- Iterate over each row of acquired data and save it to lists for conversion to a DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### response_clean()\n",
    "\n",
    "This algorthim converts the GET request response from **data_retrieval()** to a pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def response_clean(resp,UUID,tum_dict,cpm):\n",
    "    \"\"\" cleans the response from the GDC API\n",
    "    \n",
    "    Parameters:\n",
    "        resp: response from the data endpoint of the GDC API\n",
    "        UUID: UUID to use\n",
    "        tum_dict : dictionary of UUID and sample type\n",
    "        cpm: boolean; if want raw reads input False, else input True\n",
    "        \n",
    "        return clean pandas dataframe\n",
    "    \"\"\"\n",
    "\n",
    "    name = []                                                                       # creates lists to append to\n",
    "    readCnt = []\n",
    "    readCntMil = []\n",
    "    xMap = []\n",
    "\n",
    "    b= str(resp)[2:]                                                                # Convert bytes file to string and remove..\n",
    "                                                                                    # .. hardcoding.\n",
    "\n",
    "    c=b.split('\\\\n')   # split by rows\n",
    "\n",
    "    for i in c[:-1]:                                                                # iterate over every row\n",
    "        d=i.split('\\\\t')                                                            # split by column      \n",
    "\n",
    "        name.append(d[0])                                                           # append to appropriate list\n",
    "        readCnt.append(d[1])\n",
    "        readCntMil.append(d[2])\n",
    "        xMap.append(d[3])\n",
    "\n",
    "    if cpm:\n",
    "        df_dict= {'miRNA_id':name[1:],\n",
    "                  tum_dict[UUID] + \"_\" + UUID:readCntMil[1:],                                 #readCntMil or readCnt\n",
    "                 }\n",
    "        \n",
    "    else:\n",
    "        df_dict= {'miRNA_id':name[1:],\n",
    "                  tum_dict[UUID] + \"_\" + UUID:readCnt[1:],                                 #readCntMil or readCnt\n",
    "                 }\n",
    "\n",
    "    df = pd.DataFrame(data=df_dict)\n",
    " \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### data_total()\n",
    "\n",
    "This algorthm combines **file_Dict()** and **response_Clean()** by itterating over every file UUID and creating a dataframe that consists of all the file information acquired."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def data_total(tum_dict,cpm):\n",
    "    \"\"\" aggregates all data together\n",
    "    \n",
    "    parameters:\n",
    "        tum_dict: dictionary created form file_Dict()\n",
    "        cpm: if want cpm input True, if want raw input False\n",
    "        \n",
    "        returns: master Df\n",
    "    \"\"\"\n",
    "    \n",
    "    df=pd.DataFrame(columns=['miRNA_id'])\n",
    "    \n",
    "    i=0\n",
    "    \n",
    "    for UUID, site in tum_dict.items():\n",
    "        content = data_retrieval(UUID)\n",
    "        \n",
    "        temp_df = response_clean(content,UUID,tum_dict,cpm)\n",
    "        \n",
    "        df = pd.merge(df,temp_df,how='outer',on='miRNA_id')\n",
    "        i+=1\n",
    "        if i%100 == 0:\n",
    "            print(i)\n",
    "        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### main()\n",
    "\n",
    "This is that final algorithm that aggregates all the code into one run-able cell that inputs what project(s) is/are to be queried and if the values should be raw or converted to cpm. The previous cells must be run for **main()** to function properly.\n",
    "\n",
    "**Note** This cell may take some time to run. Built in it will print how many files have been aggregated every 100 files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n"
     ]
    }
   ],
   "source": [
    "def main(projects,cpm):\n",
    "    \"\"\" Uses all of the functions previously made\n",
    "    \n",
    "    Parameters:\n",
    "        projects: name(s) of the projects interested, each name coded as a string in a comma seperated list\n",
    "        cpm: if want cpm input True, if want raw input False\n",
    "    \n",
    "    Returns: usable dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    cats = file_filt(projects)\n",
    "    \n",
    "    site_dict = file_Dict(cats)\n",
    "    \n",
    "    master_df = data_total(site_dict,cpm)\n",
    "    \n",
    "    \n",
    "    \n",
    "    master_df=master_df.set_index('miRNA_id')\n",
    "    \n",
    "    finalT = master_df.transpose()\n",
    "    \n",
    "    finalT.index.name=\"UUID\"\n",
    "    \n",
    "    finalT['Site'] = \"\"\n",
    "    \n",
    "    for i in finalT.index.values:\n",
    "        if i[:13] == \"Primary Tumor\":\n",
    "            finalT.loc[i,\"Site\"] = \"Tumor\"\n",
    "\n",
    "        elif i[:13] == \"Solid Tissue \":\n",
    "            finalT.loc[i,\"Site\"] = \"Normal\"\n",
    "\n",
    "        else:\n",
    "            finalT.loc[i,\"Site\"] = \"NA\"\n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "    return finalT\n",
    "\n",
    "final = main([\"TCGA-PRAD\"],True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "final.to_csv(\"..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UUID\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>miRNA_id</th>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <th>hsa-let-7b</th>\n",
       "      <th>hsa-let-7c</th>\n",
       "      <th>hsa-let-7d</th>\n",
       "      <th>hsa-let-7e</th>\n",
       "      <th>hsa-let-7f-1</th>\n",
       "      <th>hsa-let-7f-2</th>\n",
       "      <th>hsa-let-7g</th>\n",
       "      <th>...</th>\n",
       "      <th>hsa-mir-942</th>\n",
       "      <th>hsa-mir-943</th>\n",
       "      <th>hsa-mir-944</th>\n",
       "      <th>hsa-mir-95</th>\n",
       "      <th>hsa-mir-9500</th>\n",
       "      <th>hsa-mir-96</th>\n",
       "      <th>hsa-mir-98</th>\n",
       "      <th>hsa-mir-99a</th>\n",
       "      <th>hsa-mir-99b</th>\n",
       "      <th>Site</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UUID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3</th>\n",
       "      <td>7699.123918</td>\n",
       "      <td>7683.570959</td>\n",
       "      <td>7695.849611</td>\n",
       "      <td>4706.407205</td>\n",
       "      <td>4285.454095</td>\n",
       "      <td>209.351012</td>\n",
       "      <td>599.607492</td>\n",
       "      <td>4641.330351</td>\n",
       "      <td>4683.282411</td>\n",
       "      <td>271.767491</td>\n",
       "      <td>...</td>\n",
       "      <td>1.637154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.343970</td>\n",
       "      <td>0.613933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.962247</td>\n",
       "      <td>24.966592</td>\n",
       "      <td>990.682550</td>\n",
       "      <td>12480.021610</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5</th>\n",
       "      <td>9047.765059</td>\n",
       "      <td>9068.904198</td>\n",
       "      <td>9091.965077</td>\n",
       "      <td>7124.103402</td>\n",
       "      <td>5510.482451</td>\n",
       "      <td>374.952811</td>\n",
       "      <td>1156.460377</td>\n",
       "      <td>5329.838898</td>\n",
       "      <td>5477.599345</td>\n",
       "      <td>613.889140</td>\n",
       "      <td>...</td>\n",
       "      <td>1.067633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.562320</td>\n",
       "      <td>1.281160</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>32.242525</td>\n",
       "      <td>35.872478</td>\n",
       "      <td>1782.093484</td>\n",
       "      <td>20274.569658</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4</th>\n",
       "      <td>4307.509537</td>\n",
       "      <td>4273.954612</td>\n",
       "      <td>4354.930134</td>\n",
       "      <td>5805.002069</td>\n",
       "      <td>3563.477599</td>\n",
       "      <td>508.592834</td>\n",
       "      <td>583.744774</td>\n",
       "      <td>1491.114323</td>\n",
       "      <td>1540.476114</td>\n",
       "      <td>351.078805</td>\n",
       "      <td>...</td>\n",
       "      <td>2.218507</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.109254</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>48.252537</td>\n",
       "      <td>18.302687</td>\n",
       "      <td>1915.403874</td>\n",
       "      <td>20693.128284</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce</th>\n",
       "      <td>11755.617135</td>\n",
       "      <td>11870.189263</td>\n",
       "      <td>11826.427879</td>\n",
       "      <td>6977.356413</td>\n",
       "      <td>9305.668788</td>\n",
       "      <td>395.403056</td>\n",
       "      <td>1281.484937</td>\n",
       "      <td>7355.358289</td>\n",
       "      <td>7755.240856</td>\n",
       "      <td>805.795248</td>\n",
       "      <td>...</td>\n",
       "      <td>2.239756</td>\n",
       "      <td>0.172289</td>\n",
       "      <td>0.861445</td>\n",
       "      <td>2.756623</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>47.207162</td>\n",
       "      <td>45.656562</td>\n",
       "      <td>2506.631400</td>\n",
       "      <td>15278.580836</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d</th>\n",
       "      <td>17415.647471</td>\n",
       "      <td>17405.199435</td>\n",
       "      <td>17410.116158</td>\n",
       "      <td>7923.913712</td>\n",
       "      <td>2836.949173</td>\n",
       "      <td>357.077008</td>\n",
       "      <td>758.404523</td>\n",
       "      <td>5124.454551</td>\n",
       "      <td>5563.886670</td>\n",
       "      <td>558.662651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.458362</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>46.708869</td>\n",
       "      <td>22.125254</td>\n",
       "      <td>1525.413312</td>\n",
       "      <td>14008.358429</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1882 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "miRNA_id                                            hsa-let-7a-1  \\\n",
       "UUID                                                               \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3   7699.123918   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5   9047.765059   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4   4307.509537   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  11755.617135   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  17415.647471   \n",
       "\n",
       "miRNA_id                                            hsa-let-7a-2  \\\n",
       "UUID                                                               \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3   7683.570959   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5   9068.904198   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4   4273.954612   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  11870.189263   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  17405.199435   \n",
       "\n",
       "miRNA_id                                            hsa-let-7a-3   hsa-let-7b  \\\n",
       "UUID                                                                            \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3   7695.849611  4706.407205   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5   9091.965077  7124.103402   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4   4354.930134  5805.002069   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  11826.427879  6977.356413   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  17410.116158  7923.913712   \n",
       "\n",
       "miRNA_id                                             hsa-let-7c  hsa-let-7d  \\\n",
       "UUID                                                                          \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3  4285.454095  209.351012   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  5510.482451  374.952811   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4  3563.477599  508.592834   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  9305.668788  395.403056   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  2836.949173  357.077008   \n",
       "\n",
       "miRNA_id                                             hsa-let-7e hsa-let-7f-1  \\\n",
       "UUID                                                                           \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3   599.607492  4641.330351   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  1156.460377  5329.838898   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4   583.744774  1491.114323   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  1281.484937  7355.358289   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d   758.404523  5124.454551   \n",
       "\n",
       "miRNA_id                                           hsa-let-7f-2  hsa-let-7g  \\\n",
       "UUID                                                                          \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3  4683.282411  271.767491   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  5477.599345  613.889140   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4  1540.476114  351.078805   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  7755.240856  805.795248   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  5563.886670  558.662651   \n",
       "\n",
       "miRNA_id                                            ...   hsa-mir-942  \\\n",
       "UUID                                                ...                 \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3  ...      1.637154   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  ...      1.067633   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4  ...      2.218507   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  ...      2.239756   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  ...      0.000000   \n",
       "\n",
       "miRNA_id                                           hsa-mir-943 hsa-mir-944  \\\n",
       "UUID                                                                         \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3    0.000000    6.343970   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5    0.000000    2.562320   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4    0.000000    0.000000   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce    0.172289    0.861445   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d    0.000000    0.000000   \n",
       "\n",
       "miRNA_id                                           hsa-mir-95 hsa-mir-9500  \\\n",
       "UUID                                                                         \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3   0.613933     0.000000   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5   1.281160     0.000000   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4   1.109254     0.000000   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce   2.756623     0.000000   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d   2.458362     0.000000   \n",
       "\n",
       "miRNA_id                                           hsa-mir-96 hsa-mir-98  \\\n",
       "UUID                                                                       \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3  15.962247  24.966592   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  32.242525  35.872478   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4  48.252537  18.302687   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  47.207162  45.656562   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  46.708869  22.125254   \n",
       "\n",
       "miRNA_id                                            hsa-mir-99a   hsa-mir-99b  \\\n",
       "UUID                                                                            \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3   990.682550  12480.021610   \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  1782.093484  20274.569658   \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4  1915.403874  20693.128284   \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  2506.631400  15278.580836   \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  1525.413312  14008.358429   \n",
       "\n",
       "miRNA_id                                             Site  \n",
       "UUID                                                       \n",
       "Primary Tumor_f11e4a51-f661-4357-a6ea-bcac35803ff3  Tumor  \n",
       "Primary Tumor_860889dd-ad00-411b-8eee-166d01e22fb5  Tumor  \n",
       "Primary Tumor_1cc9ab15-5258-4c8b-b1b2-2519bdf5bcd4  Tumor  \n",
       "Primary Tumor_5d7f0a0d-ffc9-42f4-b80f-a0ff244ab2ce  Tumor  \n",
       "Primary Tumor_99e05909-2544-4b6c-9571-6f299385747d  Tumor  \n",
       "\n",
       "[5 rows x 1882 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
