---
title: "Differential Analysis - BLCA (Bladder)"
author: "Conor Loy"
date: "May 25, 2018"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### This R Markdown file performs miRNA differential expression analysis for the Prostate Tissue Samples queried from the GDC API. Expression analysis is done with the edgeR and limma packages.


## Load Packages and Read in Data


##### First, all packages and data must be read loaded. edgeR and limma take raw count reads, thus this markdown file will not be using cpm (count per million) values queried from the GDC API.

```{r,echo=TRUE,warning=FALSE}
library(knitr)
library(rmarkdown)
library(dplyr)
library(ggplot2)
library(ggthemes)
library(edgeR)
library(limma)



raw_rd <- read.csv("../../CleanData/PRAD_rds.csv")

head(raw_rd[1:5,1:5])
```


## Prepare Data for Differential Expression Analysis

#### Create tumor and normal data sets from the non-normalized data

##### Filter by place-holder variable "Site" to create two seperate data sets for organized merge later. This is why the data set is configured how it is upstream.


```{r}
tumor_rd<-raw_rd%>%
  filter(Site == "Tumor")

normal_rd<-raw_rd%>%
  filter(Site == "Normal")
```



#### Transpose all data sets for proper configuration for edgeR DGEList

##### Now that the data sets are seperated by sample site, the data can be transposed for creation of a DGEList. A DGEList object has requires a matrix or dataframe with columns representing a sample and each row a microRNA.


```{r}
# Start with the master
raw_t <-t(
    subset(raw_rd,select=-c(Site))
    )
  
colnames(raw_t) <- raw_t[1,]
raw_t <- raw_t[-1,]

raw_t <- as.data.frame(raw_t)


#Also do the Tumor
tumor_t <- t(
    subset(tumor_rd,select=-c(Site))
    )
  
colnames(tumor_t) <- tumor_t[1,]
tumor_t <- tumor_t[-1,]

tumor_t <- as.data.frame(tumor_t)


# and Normal 
normal_t <- t(
    subset(normal_rd,select=-c(Site))
    )

colnames(normal_t) <- normal_t[1,]
normal_t <- normal_t[-1,]
normal_t <- as.data.frame(normal_t)



head(raw_t[1:5,1:2])
```


#### Grouping the different samples

##### Here we create a list that contains T's and N's. This list will be used later in the design matrix for modeling. The list contains first the letter T itterated over the number of Tumor-Tissue samples in the dataframe. Next, the list contains the letter N itterated over the number of Normal-Tissue samples in the dataframe.


```{r}
tumorSamp=c("T")

tumorSamp=rep(tumorSamp,length(colnames(tumor_t)))

nrmSamp=c("N")

nrmSamp=rep(nrmSamp,length(colnames(normal_t)))

groups_samp <- append(tumorSamp,nrmSamp)
```


#### Create Master dataframe in order by group to create DGEList

##### The transposed data frames are then merged so the Tumor samples are the first columns and the Normal samples are the next columns. This will allow it to match the groups_samp object we created for experiment design.


```{r}
mstr_ordered <- merge(tumor_t,normal_t, by.x=0, by.y=0,all=TRUE)

rownames(mstr_ordered) <- mstr_ordered$Row.names

mstr_ordered <- subset(mstr_ordered,select = -c(Row.names))

mstr_ordered[] <- lapply(mstr_ordered, function(x) as.numeric(as.character(x)))

head(mstr_ordered[1:5,1:2])
```


#### filter out lowly expressed miRNA

##### Remove miRNA that are found at very low rates. This allows for more meaningful adjusted P-values.


```{r}
cpms<-cpm(mstr_ordered)
keep <- rowSums(cpms>1)>=100## at least 4 replicates
counts <- mstr_ordered[keep,]
```


#### Create DGEList object from Master Dataframe

##### Now that the data has been seperated, transposed, merged and filtered it is in the proper form for conversion to a DGEList object. The code below creates the DGEList Object and assigns Site-type to each sample with the "group" argument.


```{r}
dge <- DGEList(counts = counts, group=groups_samp,remove.zeros=TRUE)
```


#### Create Normalization Factors

##### Each sample is from a different individual and site, thus each library needs to normalized based on size and distribution. Here factors are assigned with the "calcNormFactors()" function using the upper quartile method. This method calculates scale factors from the 75% upper quantile of the counts in each library.  

##### This is done to account for the RNA composition, if a small group of RNA are highly over expressed in one sample type it will throw off the composition of everything else. A normalization factor less than 1 indicates a small number of high count genes are monoppolizing the sequencing.
  

```{r}
y <- calcNormFactors(dge,method = "upperquartile")
```


#### Unsupervised Clustering of Groups 

##### Here a unsupervised program produces a MDS (multi-dimentional scaling) plot to explore the data. Seperation of the experimental groups indicates a difference in the groups expression levels.


```{r}
lcpm <- cpm(y,log=TRUE)
par(mfrow=c(1,2))
group <- groups_samp
col.group <- groups_samp

mdsPlot <- plotMDS(lcpm,labels=group, col=ifelse(col.group=="T","red","blue"))
```


##### The plot looks good! The grouping of normal tissue samples in the upper right corner indicates similarity amongst normal samples.


## Differential Analysis 

#### Design Matrix

##### To start the differntial Analysis a design matrix must first be created to indicate experimental design. This allows the model to compare groups of interest.


```{r}
design <- model.matrix(~factor(groups_samp))
#colnames(design) <- c("tumor","normal")
#contrast <- makeContrasts(tumor- normal, levels = design)
head(design)
```


#### Analysis

##### With the DGEList and the design matrix created we can run our statistics. Both graphics produced indicate proper filtering and modeling. The below table are the top 10 differently regulated miRNA.


```{r}
par(mfrow=c(1,2))

v <- voom(y, design, plot=TRUE, normalize="quantile")
fit <- lmFit(v,design)

#fit <- contrasts.fit(fit, contrast)

fit <- eBayes(fit)

plotSA(fit) # "plot log residual standard deviation versus average log expression for a fitted microarray linear model"

topTable(fit, coef =2)
```



```{r}
summary(decideTests(fit))


temp <- topTable(fit, coef=2, n=Inf)

write.csv(temp, "../../CleanData/PRAD_DifExp_outPut.csv")
```